[SYCL] Document the SYCL Runtime Plugin Interface. (#206)
intel/llvm: sycl/doc/SYCLPluginInterface.md

# The SYCL Runtime Plugin Interface.

## Overview
The SYCL Runtime Plugin Interface (PI) is the interface layer between
device-agnostic part of the SYCL runtime and the device-specific runtime layers

> which control execution on the device. It employs the “plugin” mechanism to bind

Reviewer: on "the device" -> on devices.

to the device specific runtime layers similarly to what is used by libomptarget
or OpenCL.

> ![PI in SYCL runtime architecture](images/SYCL_RT_arch.svg)

Author: ok, will update

The plugin interface and the discovery process behind it allows to dynamically
plug in implementations based on OpenCL and �native� runtime for a particular
device � such as OpenCL for

> FPGA devices or native runtmies for GPUs. Implementations of the PI are

Reviewer 2: ```suggestion
FPGA devices or native runtimes for GPUs. Implementations of the PI are
```

�plugins� - dynamic libraries or shared objects which expose a number of entry
points implementing the PI interface. The SYCL runtime collects those function

> pointers into a the PI interface dispatch table - one per plugin - and uses this

Reviewer 2: ```suggestion
pointers into a PI interface dispatch table - one per plugin - and uses this
```

table to dispatch to the device(s) covered by the corresponding plugin. Further
sections give complete descriptions of all entry points as well as the dispatch
table.

> 

Reviewer: This document does not five complete description if all PI interfaces for now. So I suggest just to remove this last sentence.

Author: ok

PI is based on a subset of OpenCL 1.2 runtime specification, it follows its

> platform, excution and memory models in all aspects except those explicitly

Reviewer 2: ```suggestion
platform, execution and memory models in all aspects except those explicitly
```

mentioned in this document. A part of PI API types and functions have exact
matches in OpenCL. Whenever there is such a match, the semantics also fully
matches unless the differences are explicitly specified in this document. While
PI has roots in OpenCL, it does have many differences, and the gap is likely
to grow, for example in the areas of memory model and management, program

> management.

Reviewer 3: After the (welcomed) latest changes to the naming scheme, is not so clear anymore which functions map 1:1 to OpenCL. 
Maybe the document could contain the mapping down below?

Author: I'd expect (at least part of) PI<->OpenCL mapping is to evolve, so I suggest that instead every function in the pi.h is commented whether it has direct OpenCL couterpart or not.

## Discovery and linkage of PI implementations

![PI implementation discovery](images/SYCL_plugin_discovery.svg)

Device discovery phase enumerates all available devices and their features by
querying underlying plugins found in the system. This process is only performed
once before any actual offload is attempted.

### Plugin discovery

Plugins are physically dynamic libraries stored somewhere in the system where
the SYCL runtime runs. TBD - describe the process in details.

#### Plugin binary interface
TBD - list and describe all the symbols plugin must export in order to be picked
up by the SYCL runtime for offload.

#### OpenCL plugin

OpenCL plugin is a usual plugin from SYCL runtime standpoint, but its loading
and initialization involves a nested discovery process which finds out available
OpenCL implementations. They can be installed either in the standard Khronos
ICD-compatible way or not, and the OpenCL plugin can hook up with both.

> 

Reviewer: It sounds that the "OpenCL plugin" would support ANY (not "both") OpenCL implementation, no matter how that is installed, which can't be true. Maybe elaborate on what exactly you meant, or just add TBD comment.

Reviewer: Oh, I see you describe "both" ways below, which sounds fine for now.

##### Linking to ICD-compatible OpenCL installations

The plugin dynamically loads the OCL ICD loader - libOpenCL.so from system
paths, and on success it asks the loader to provide all OpenCL implementations
installed on the system. To do that it finds and calls the "clGetPlatformIDs"
in the ICD loader and records the returned platform objects for later use. The
plugin may choose to dynamically link with libOpenCL.so at build time.
 
##### Linking to ICD-incompatible OpenCL installations

For the OpenCL plugin to hook up such installation, it must be described in a
configuration file similarly to plugin configuration file. *TBD - location and
format of custom OpenCL installation config file*. The plugin will dynamically
load the dynamic library pointed to in the configuration file and query it for
OpenCL API function addresses similarly ICD loader. In more details, it will:
- search for  the ```clGetExtensionFunctionAddress``` function in the 
  library and call it with the string *“clIcdGetPlatformIDsKHR”* to find
  the ```clIcdGetPlatformIDsKHR``` function pointer

> - then, using ```clIcdGetPlatformIDsKHR```, find the set of

Reviewer 3: From the text above, seems to refer to an OpenCL implementation that has the `cl_khr_icd` extension but is not registered in `/etc/OpenCL/vendors` or wherever this happens on Windows. 
Seems to me that, for it to be really flexible, it shouldn't rely on the cl_khr_icd extension, and simply work with any libOpenCL.so that exposes the `clGetPlatforms` function.

> - then, using ```clIcdGetPlatformIDsKHR```, find the set of

Author: ICD-incompatible runtimes are those like packaged with SYCL distribution but not installed in such a way that they can be found by the standard ICD loader.

> Seems to me that, for it to be really flexible, it shouldn't rely on the cl_khr_icd extension, and simply work with any libOpenCL.so that exposes the clGetPlatforms function.

hope we can we change this later

  ```cl_platform_id``` objects that the library provides, and store them
  internally to enumerate available devices when required

### Device enumeration by plugins

TBD

## PI API Specification

PI allows for two levels of API implementation - basic and extended. The basic
level comprises core APIs needed to perform SYCL offload. The extended level is

> needed only for OpenCL interoperability. 

Reviewer 4: Just OpenCL or "device-specific runtime"?

Reviewer: Let's follow the notation used in naming of PI interfaces (#222 and #259). There are core "pi" functionality, interoperability "picl" interfaces, and extended ones, those that are not in OpenCL but needed for SYCL.

Reviewer: > Just OpenCL or "device-specific runtime"?

I think it can be later in the long term where we'd want better interop with other runtimes, but today it is OpenCL only. The other runtimes interop PI interfaces would probably be prefixed with "piext".

### The basic PI APIs

This level requires availability of functions representing core functionality,
such as device memory management, kernel creation and parameter setting,
enqueuing kernel for execution, etc. There are functions which fully match
semantics of the corresponding OpenCL X.X functions, as well as functions which
don't have direct mapping to OpenCL. See the
[pi.h](../include/CL/sycl/detail/pi.h) header for the full list and
descriptions of PI APIs. [TBD: link to pi.h doxygen here]

*__Examples of PI OpenCL-based APIs__*

    piCreateKernel
    piReleaseKernel
    piSetKernelArg

*__Example of PI non OpenCL-based APIs__*

Selects the most appropriate device image based on runtime
information and the image characteristics
```

> cl_int piSelectDeviceImage(pi_context ctx, pi_device_image **images,

Reviewer: This better be updated with the new PI interface names in #259

> cl_int piSelectDeviceImage(pi_context ctx, pi_device_image **images,

Author: > non OpenCL-based API is using cl_int, which is OpenCL-based type. Is it a bug?

Yes, thanks.

> This better be updated with the new PI interface names in #259

OK

                             cl_uint num_images,
                             pi_device_image **selected_image);
```

PI also defines few types and string tags to describe a device binary image.
Those are used to communicate to plugins information about the images where it

> is needed, currently only in the ```piSelectDeviceImage``` function. The main

Reviewer 4: > PI also defines few types and string tags to describe a device binary image. Those are used ... in the ```piSelectDeviceImage``` function.

I see `pi_device_image` type in `piSelectDeviceImage` declaration, but what is _**string tags**_?

Author: @bader,

/// Target identification strings for
/// pi_device_binary_struct.DeviceTargetSpec
///
#define PI_DEVICE_BINARY_TARGET_UNKNOWN "<unknown>"
#define PI_DEVICE_BINARY_TARGET_SPIRV32 "spir"
#define PI_DEVICE_BINARY_TARGET_SPIRV64 "spir64";

type is ```pi_device_image```, whose detailed description can also be found in

> the header.  The layout of this type strictly match layout of the coresponding

Reviewer 2: ```suggestion
the header.  The layout of this type strictly matches the layout of the corresponding
```

device image descriptor type defined in the ```clang-offload-wrapper``` tool

> which wrapps device binaries into a host object for further linkage. The wrapped

Reviewer 2: ```suggestion
which wraps device binaries into a host object for further linkage. The wrapped
```

images reside inside this descriptor in a data section.

### The Extended PI APIs

> These are APIs needed to implement SYCL-to-OpenCL interoperability. Only

Reviewer 4: I think this section should be generalized to describe extended PI APIs as a mechanism to provide interoperability with device specific API not **only OpenCL**.
SYCL should be able to interoperate with OpenMP if SYCL runtime uses extended OpenMP plug-in API. Right?

Or this section should be called "The extended OpenCL plugin API".

Reviewer: > Or this section should be called "The extended OpenCL plugin API".

+1

Author: > with device specific API not only OpenCL.
this was specifically for OpenCL interop. Interop with any device is a big separate step

the OpenCL plugin will provide these APIs, generic plugins won't, because
SYCL runtime will cast PI object pointers to/from OpenCL object pointers when
using result of/invoking the extended APIs in interoperability scenario. These
APIs match semantics of the corresponding OpenCL APIs exactly.

> Full list The list of such APIs can be found in the

Reviewer 2: ```suggestion
The full list of such APIs can be found in the
```

[pi.h](../include/CL/sycl/detail/pi.h) header. One example is:

>     piOpenCLCreateProgramWithSource

Reviewer: piclProgramCreate...

Author: ok

 
### Extension mechanism

This section describes a mechanism for SYCL or other runtimes to detect
availability of and obtain interfaces beyond those defined by the PI dispatch.

> The ```piGetExtensionFunctionAddressForPlatform``` function can be used

Reviewer: Why can't we just define more PI interfaces and make them optional to implement?  Unlike OpenCL, the PI API is not a standard.

Reviewer 3: That makes sense, but also means some kind of agreement for everyone using the PI interface will need to be in place as to what is an extension to PI and what is a core PI functionality - hence why I suggested in a different place to at least have a version number.
If we take what is in this repo SYCL branch as the golden PI interface, pull requests can be used to add new functionality.

Author: OK, I'll remove this one leaving TBD for the extension mechanism in this place. Originally PI was design to be binary compatible with OpenCL so that OpenCL dispatch tables could be used in place of PI dispatch table, now the approach has evolved.

to find a specific function pointer and call it to get an additional table of
function pointers (if they are provided by implementation), for example:

    getOMPDispatchTable =
      (getOMPDispatchTable_type)
      piGetExtensionFunctionAddressForPlatform(platf_id,_
                                         “getOMPDispatchTable”);
    OMPDispatchTable = getOMPDispatchTable(OMPVersion);

> ### Plugin access to SYCL runtime interface

Reviewer: This section is very raw in my opinion, and should all be TBD maybe just discussing the possibility if that would be needed. Again, I think  we would better extend PI API (e.g. some callback style) when/if we need more handshake between SYCL RT and underlying plugins.

Author: OK

The PI design allows plugins to get access to some of the SYCL runtime
functionality, such as global resource acquisition and release, asynchronous
completion signaling (TBD). The same approach as for obtaining additional
interfaces can be used here. For example:

Types defined by PI:

    enum ErrorCode {…};
    …
    struct PIDisptachTable {
      CFunctionPtrType1 function1;
      CFunctionPtrType2 function2;
    …
    };
    typedef ErrorCode

>       (*PIDisptachTableSetterF)(PIDisptachTable *sycl_rt_intf);

Reviewer 3: Disptach -> Dispatch ?

Author: OK

SYCL RT code:

    static PIDisptachTable sycl_rt_intf;
    …
    PIDisptachTableSetterF f =
      (SYCLRtInterfaceSetterF)clGetExtensionFunctionAddressForPlatform(
        platf_id, “setSYCLRtInterface”);

>     if (f) f(&sycl_rt_intf);

Reviewer 3: The example is not very clear, what is it trying to do? and what is "setSYCLRtInterface" doing?
As @smaslov-intel  suggests above, its better to leave this as a TBD, and after there is more implementation experience, the documentation can be updated - this shouldn't be a blocker for this PR to come through.

Author: OK

  internally to enumerate available devices when required

### Device enumeration by plugins

TBD

## PI API Specification

PI allows for two levels of API implementation - basic and extended. The basic
level comprises core APIs needed to perform SYCL offload. The extended level is
needed only for OpenCL interoperability. 

### The basic PI APIs

This level requires availability of functions representing core functionality,
such as device memory management, kernel creation and parameter setting,
enqueuing kernel for execution, etc. There are functions which fully match
semantics of the corresponding OpenCL X.X functions, as well as functions which
don't have direct mapping to OpenCL. The two lists are below.
 
*__OpenCL-based APIs__*

    piGetPlatformIDs
    piGetDeviceIDs
    piCreateContext
    piReleaseContext
    piCreateCommandQueue
    piReleaseCommandQueue
    piCreateBuffer
    piReleaseMemObject
    piReleaseSampler
    piReleaseProgram
    piBuildProgram
    piCreateKernel
    piReleaseKernel
    piSetKernelArg
    piWaitForEvents
    piReleaseEvent
    piGetEventProfilingInfo
    piFlush
    piFinish

>     piEnqueueReadBuffer

Reviewer 5: I could be wrong or misunderstanding it, but I think this list is missing some of the calls from the pi.h interface (it may be intentional though as its the minimal subset). For example piEnqueueReadBufferRect and piEnqueueWriteBufferRect which are used for handling 2d/3d buffer transfers.

Author: Should we maybe avoid this duplication then and maintain APIs in one place - pi.h?

Reviewer 5: I think just directing people to the relevant place to find the plugin interface (pi.h) and giving an example interface call would work, so they know what they're looking for. It means you wouldn't have to maintain the list in the .md every time pi.h changed.

Reviewer 3: Doxygen or similar could be used to generate a more readable documentation directly from the headers

Author: > It means you wouldn't have to maintain the list in the .md every time pi.h changed.

Right, that's what I meant by avoiding duplication. Will remove most of the APIs and leave just few examples.

> Doxygen or similar could be used to generate a more readable documentation directly from the headers

Good idea. Should be doxygen, I think, to be compatible with llvm/clang. We can start with maintaining doxygen in the pi.h
